rbd: detect migrated volhandle in DeleteVolume and delete rbd image

humblec opened the pull request:
rbd: detect migration volID in DeleteVolume() and delete rbd image
    
    This commit adds the logic to detect a passed in volumeID
    is a migrated volume ID and if yes, the driver connect to the
    backend cluster and clean/delete the image. The logic
    only applied if its a migration volume ID. The migration volume ID
    carry the information like mons, pool and image name which is
    good enough for the driver to identify and connect to the backend
    cluster for its operations.
    
    migration volID format:
    <mig>_mons-<monsHash>_image-<imageUID>_<poolHash>
    
    Details on the hash values:
    
    * MonsHash: this carry a hash value (md5sum) which will be acted as the
    `clusterID` for the operations in this context.
    
    * ImageUID: this is the unique UUID generated by kubernetes for the created
    volume.
    
    * PoolHash: this is an encoded string of pool name.

Updates #2509 
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
 	return &csi.DeleteVolumeResponse{}, nil
 }
 
+// isMigrationVolHandle check the passed in volHash is a volume handle
```
move this to a new file called migration.go?

humblec replied:
Done.. thanks.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+		return true
+	}
+
+	return false
```
```suggestion
return strings.Contains(volHash, migVersion) && strings.Contains(volHash, migImageNamePfx)
```

humblec replied:
Done.. thanks

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+
+	// Parse migration version, mons and image
+	for _, id := range handSlice[:3] {
+		fmt.Println(id)
```
remove println

humblec replied:
Done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+// parseMigrationVolHandle decode the volume handle and generate a migrationHandle
+// struct which consists of version, mon, image name and pool information.
+func parseMigrationVolHandle(vh string) (*migrationHandle, error) {
+	mh := new(migrationHandle)
```
mh := &migrationHandle{}

humblec replied:
done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+	}
+
+	// Store pool
+	mh.Pool = strings.Join(handSlice[3:], migHandleFieldSep)
```
define a const for `3`

humblec replied:
Done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+	mh.Pool = strings.Join(handSlice[3:], migHandleFieldSep)
+
+	// Parse migration version, mons and image
+	for _, id := range handSlice[:3] {
```
`id` is not a perfect name may be something else?

humblec replied:
changed to `field` .. done

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
 
+	if isMigrationVolHandle(volumeID) {
+		log.DebugLog(ctx, "migration volume handle : %s", volumeID)
+		pMH, pErr := parseMigrationVolHandle(volumeID)
```
`pMH` can be `parsedMigHandle`?

Yuggupta27 replied:
+1

humblec replied:
done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+		log.DebugLog(ctx, "migration volume handle : %s", volumeID)
+		pMH, pErr := parseMigrationVolHandle(volumeID)
+		if pErr != nil {
+			log.DebugLog(ctx, "failed to parse migration volume handle: %s , err: %v", volumeID, pErr)
```
use ErrorLog

humblec replied:
Done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
 	}
 	defer cs.OperationLocks.ReleaseDeleteLock(volumeID)
 
+	if isMigrationVolHandle(volumeID) {
```
move this complete logic to migration.go?

humblec replied:
Not sure its required, if really needed I can do that. :)

humblec replied:
nw, I moved the logic completely to migration.go.. Thanks..

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+		monsMap["monitors"] = rv.Monitors
+		clusterID, cErr := getClusterIDFromMigrationVolume(monsMap)
+		if err != nil {
+			log.DebugLog(ctx, "failed to get ClusterID for the rbd image : %s", rv.RbdImageName)
```
Error log

humblec replied:
Done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+		rv.ClusterID = clusterID
+		err = rv.Connect(cr)
+		if err != nil {
+			log.DebugLog(ctx, "failed to get connected to the rbd image : %s", rv.RbdImageName)
```
Error log?

humblec replied:
Done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+			rv.Pool = pMH.Pool
+		}
+		log.DebugLog(ctx, "filled rbd volume: %+v", rv)
+		monsMap := make(map[string]string)
+		monsMap["monitors"] = rv.Monitors
```
why create an extra map just for this variable update `getClusterIDFromMigrationVolume` to take a string input?

humblec replied:
yeah, was thinking about it.. I dont see a strong reason to put that on map, so let me revert that to string.

humblec replied:
done.. thanks.. now it take string input.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+		log.DebugLog(ctx, "filled rbd volume: %+v", rv)
+		monsMap := make(map[string]string)
+		monsMap["monitors"] = rv.Monitors
+		clusterID, cErr := getClusterIDFromMigrationVolume(monsMap)
```
why do you need clusterID here?

humblec replied:
`deleteImage->..->addRbdManagerTask()` use the `clusterID`.

Madhu-1 replied:
still, I don't see that code where we are using the clusterID. can you please point me to the code where the clusterID is getting used?

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+		// if connected , delete it
+		err = deleteImage(ctx, rv, cr)
+		if err != nil {
+			log.DebugLog(ctx, ",failed to delete rv: %+v", rv)
```
Error log

humblec replied:
Done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+		}
+		defer rv.Destroy()
+		// if connected , delete it
+		err = deleteImage(ctx, rv, cr)
```
check image exists before deleting need to handle not found error

humblec replied:
`deleteImage` try to get image and return error, so its handled already?

Madhu-1 replied:
deleteImage internally calls getImageID first and its returns the `ErrImageNotFound` error as we don't have a check for that here it always returns an internal server error.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -77,6 +77,13 @@ const (
 	intreeMigrationKey   = "migration"
 	intreeMigrationLabel = "true"
+
+	// these are the migration volume handle identifiers.
+	migHandleLength   = 4
+	migHandleFieldSep = "_"
+	migVersion        = "migrate-alpha1"
+	migImageNamePfx   = "kubernetes-dynamic-pvc-"
+	migMonPfx         = "mons-"
```
are these are kubernetes defined one?  please use full names `migMonPfx` to `migratedMonitorPrefix` etc

Madhu-1 replied:
Do we have any documentation for these consts in kubernetes?

humblec replied:
Yep, added details. done.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +878,48 @@ func (cs *ControllerServer) DeleteVolume(
+			return nil, pErr
+		}
+		rv := &rbdVolume{}
+		if pMH != nil {
```
why this check is needed? in any case both error and nil values are returned from parseMigrationVolHandle?

humblec replied:
placed it for more safety as there is a dereferencing in place, but removed it in current refresh of the patchset.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+	for _, id := range handSlice[:3] {
+		fmt.Println(id)
+		switch {
+		case id == migVersion:
+			mh.Version = id
```
what are we doing with this Version?

humblec replied:
this version helps to identify the handle version wrt to the migration feature state, like `alpha`.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+		case strings.Contains(id, migImageNamePfx):
+			mh.ImageName = id
+		case strings.Contains(id, migMonPfx):
+			newMons := strings.Trim(id, migMonPfx)
+			mh.Mons = newMons
```
why create a new extra variable? direct use strings.Trim

humblec replied:
removed it. Done.

Madhu-1 commented on internal/rbd/controllerserver_test.go:
```diff
@@ -67,3 +68,146 @@ func TestIsThickProvisionRequest(t *testing.T) {
+
+func TestIsMigrationVolHandle(t *testing.T) {
+	t.Parallel()
+	req := &csi.DeleteVolumeRequest{
+		VolumeId: "fake",
+	}
```
why this is needed? use can use a variable?

humblec replied:
any wrong entry is good enough, just put it as a variable for now.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -175,6 +182,15 @@ type imageFeature struct {
+	Version   string
+	Mons      string
+	ImageName string
+	Pool      string
```
is it pool Name or Pool ID?

humblec replied:
PoolName, renamed the field for more clarity

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -175,6 +182,15 @@ type imageFeature struct {
+// from migrated volume handle.
+type migrationHandle struct {
+	Version   string
+	Mons      string
```
use Monitors

humblec replied:
Done.

Madhu-1 reviewed (dismissed):
@humblec is this PR tested already?

Please add an E2E testing

Yuggupta27 commented on internal/rbd/rbd_util.go:
```diff
@@ -77,6 +77,13 @@ const (
 	intreeMigrationKey   = "migration"
 	intreeMigrationLabel = "true"
+
+	// these are the migration volume handle identifiers.
```
```suggestion
	// migration volume handle identifiers.
```

humblec replied:
Done. thanks.

Yuggupta27 commented on internal/rbd/controllerserver.go:
```diff
@@ -794,6 +795,50 @@ func (cs *ControllerServer) checkErrAndUndoReserve(
+			mh.Mons = newMons
+		}
+	}
+	if mh.ImageName == "" || mh.Pool == "" || mh.Mons == "" {
```
Can we return separate warnings so that we can identify what is empty in logs?
E.g: if mons information is not present, passing "monitor information can not be empty" will give a more clear idea than a general ErrInvalidVolID

humblec replied:
Done. thanks.

humblec commented:
> @humblec is this PR tested already?

@Madhu-1 yes, I have tested this delete path and it is working.  

Reg, adding E2E, unfortunately  I dont see a straight way here, that said, the volHandle and Image name has to be manipulated either by following our normal PVC creation with proper SC, or the SC details and other attributes in the PVC/PV spec need to be changed after creation if its static PV workflow, both looks to be difficult or not possible to achieve. 

As our current E2E confirm other code paths are uneffected, may be go ahead without e2e ?

humblec commented:
@Madhu-1 @Yuggupta27 all comments are addressed except e2e https://github.com/ceph/ceph-csi/pull/2523#issuecomment-923665551. ptal.. Thanks.

Madhu-1 commented on internal/rbd/controllerserver_test.go:
```diff
@@ -67,3 +68,149 @@ func TestIsThickProvisionRequest(t *testing.T) {
+func TestIsMigrationVolHandle(t *testing.T) {
+	t.Parallel()
+	volID := "wrong_volume_handle"
+	req := &csi.DeleteVolumeRequest{
+		VolumeId: volID,
+	}
```
this is not needed. use the volID defined at line 74?

humblec replied:
@Madhu-1 we are using `volID` defined at line 74 here, I am sorry, I didnt get the comment.

Madhu-1 replied:
why create req variable and send req.VolumeID to the isMigrationVolHandle?

```
volID := "wrong_volume_handle"
if isMigrationVolHandle(volID) {
	t.Error("request is not a migration delete request")
}
```

humblec replied:
done.

Madhu-1 commented on internal/rbd/controllerserver_test.go:
```diff
@@ -67,3 +68,149 @@ func TestIsThickProvisionRequest(t *testing.T) {
+	}
+	tests := []struct {
+		name    string
+		args    args
```
use the string here? instead of a struct as it contains only one parameter.

humblec replied:
done.

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,115 @@
+	// Parse migration version, mons and image
+	for _, field := range handSlice[:migHandleSplitLength] {
+		switch {
+		case field == migVersion:
+			// version is an identifier of migration handle format tied with a
+			// migration feature version, ex: alpha
+			mh.Version = field
```
Do we need this check? what are we doing based on the Version?

humblec replied:
the version helps us to identify the vol handle. its similar to the version of vol handle we have.

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,115 @@
+	rv.Monitors = parsedMigHandle.Monitors
+	rv.Pool = parsedMigHandle.PoolName
+
+	log.DebugLog(ctx, "filled rbd volume: %+v", rv)
```
This is not needed, please remove it.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -175,6 +189,15 @@ type imageFeature struct {
+// migrationHandle is a struct which consists of required fields of a rbd volume
+// from migrated volume handle.
+type migrationHandle struct {
+	Version   string
+	Monitors  string
+	ImageName string
+	PoolName  string
```
any specific reason to export these fields in an unexported structure?

humblec replied:
unexported @Madhu-1 .. done. thanks.

Madhu-1 commented:
> > @humblec is this PR tested already?
> 
> @Madhu-1 yes, I have tested this delete path and it is working.
> 
> Reg, adding E2E, unfortunately I dont see a straight way here, that said, the volHandle and Image name has to be manipulated either by following our normal PVC creation with proper SC, or the SC details and other attributes in the PVC/PV spec need to be changed after creation if its static PV workflow, both looks to be difficult or not possible to achieve.
> 

I  think it can be done easily. Can you make use of the `getStaticPV`  function

* Create static PVC and static PV
* Create an application pod and verify you are able to get the pod running

The above steps are already done in static migration tests

extra steps are below
* when creating the static PV generate the volumeHandle use it
* Change the reclaimPolicy from `Retain` to delete which will trigger rbd volume deletion.
* 

> As our current E2E confirm other code paths are uneffected, may be go ahead without e2e ?

Madhu-1 commented:
@humblec as we are adding the monIp to the volumeHandle. In case of the mon changes in the ceph cluster what are the plans of recovery here?

humblec commented:
> @humblec as we are adding the monIp to the volumeHandle. In case of the mon changes in the ceph cluster what are the plans of recovery here?

Admin has to make sure the `mons` are part of the configmap, considering these were part of the `SC` , thats something admin has to take care of.

Madhu-1 commented:
> > @humblec as we are adding the monIp to the volumeHandle. In case of the mon changes in the ceph cluster what are the plans of recovery here?
> 
> Admin has to make sure the `mons` are part of the configmap, considering these were part of the `SC` , thats something admin has to take care of.

Am checking about the mon which is getting added to the volumeHandle. if the mon changes the rbd image cannot be deleted from the csi side.

Madhu-1 commented on internal/rbd/controllerserver_test.go:
```diff
@@ -67,3 +68,149 @@ func TestIsThickProvisionRequest(t *testing.T) {
+	}
+}
+
+func TestParseMigrationVolHandle(t *testing.T) {
```
this unittesting should be done at migration_test.go?

humblec replied:
yep, it can be moved to .

humblec commented:
> > > @humblec as we are adding the monIp to the volumeHandle. In case of the mon changes in the ceph cluster what are the plans of recovery here?
> > 
> > 
> > Admin has to make sure the `mons` are part of the configmap, considering these were part of the `SC` , thats something admin has to take care of.
> 
> Am checking about the mon which is getting added to the volumeHandle. if the mon changes the rbd image cannot be deleted from the csi side.

I get it, but again, its admins responsibility to make sure the monitors remain till the lifecycle of pvc considering those were part of the SC, iow, if we can not or kubernetes can not delete the volume due to all the monitors mentioned in SC are not available, admin has to do the cleanup manually. more or less, considering its a string slice , if any one  of it remains, its good enough or still  work. so an agreed schema and its a best effort on what kube/csi can do.

ShyamsundarR commented:
> > > > @humblec as we are adding the monIp to the volumeHandle. In case of the mon changes in the ceph cluster what are the plans of recovery here?
> > > 
> > > 
> > > Admin has to make sure the `mons` are part of the configmap, considering these were part of the `SC` , thats something admin has to take care of.
> > 
> > 
> > Am checking about the mon which is getting added to the volumeHandle. if the mon changes the rbd image cannot be deleted from the csi side.
> 
> I get it, but again, its admins responsibility to make sure the monitors remain till the lifecycle of pvc considering those were part of the SC, iow, if we can not or kubernetes can not delete the volume due to all the monitors mentioned in SC are not available, admin has to do the cleanup manually. more or less, considering its a string slice , if any one of it remains, its good enough or still work. so an agreed schema and its a best effort on what kube/csi can do.

This comment caught my curiosity, apologies for late suggestions to the PR.

So if I understand this right, the in-tree migration path has the ability to morph/create a volume handle and invoke CSI related gRPCs. IOW, pass in a volume handle crafted using the existing `volume` parameters.

Why not create a volume handle whose prefix is a hash of volumes.rbd.monitors and add to CSI ceph config this hash as a cluster ID and store new/changing MONs and other information here?

IOW, for in-tree migration to ceph-CSI, user/admin needs to update CSI config with appropriate mapping, and then open up the feature gate.

Possibly this was considered and not viable, but checking nevertheless.

humblec commented:
> 
> Why not create a volume handle whose prefix is a hash of volumes.rbd.monitors and add to CSI ceph config this hash as a cluster ID and store new/changing MONs and other information here?
> IOW, for in-tree migration to ceph-CSI, user/admin needs to update CSI config with appropriate mapping, and then open up the feature gate.
> Possibly this was considered and not viable, but checking nevertheless.
> 

Considering `monitors` and `clusterId` are the analogous ones, that was the very first thought. however  at migration time, taking out admin's burden as much as , is the expectation, with that in mind the current  form of vol handle has been considered. Also, all the mons going down is not a really an expected scenario ( if that happens ,current intree PVs are broken anyway), so this form helps to a very good extent. also, even if all goes down, as long as there is a marker of one old IP in one of the clusterid (even along with new mons), it will work. Regardless, still I think , we can ask the admin to also have config update with mon hash as clusterID as long as its not that heavy change. Calculating hash ( md5sum) of mons as a clusterID and updating config map should be possible to request to admin before he turn on the migration.  considering its a minor change here,  let me try to adjust the PR with this.

humblec commented:
@ShyamsundarR  Used `monitors` hash (md5sum of monitors) as `clusterID` and additionally  encoded `pool name` in the handle. updated unit tests as well with new hash.

note: e2e is still pending

ShyamsundarR commented:
> @ShyamsundarR Used `monitors` hash (md5sum of monitors) as `clusterID` and additionally encoded `pool name` in the handle. updated unit tests as well with new hash.
> 
> note: e2e is still pending

Just some corner cases to deal with in the migration handle design. Volumes may have varying existing MONs, and hence we may end up with multiple hashed versions of the same as the `clusterID` in the config.

humblec commented:
> > @ShyamsundarR Used `monitors` hash (md5sum of monitors) as `clusterID` and additionally encoded `pool name` in the handle. updated unit tests as well with new hash.
> > note: e2e is still pending
> 
> Just some corner cases to deal with in the migration handle design. Volumes may have varying existing MONs, and hence we may end up with multiple hashed versions of the same as the `clusterID` in the config.

This shouldnt be an issue as this vol handle is applicable only for existing PVs, so before turning on the migration, admin has to make sure there is a clusterID created for each in tree SC based on the `monitors` field. If there is no difference in the monitors field in in tree SCs, he dont want to repeat it. For all existing PVs, the monitors are already recorded  inside pv->spec based on the existing SCs , so this is actually constant and the mapping should work as long as admin has created one for each SC->monitors string.

humblec commented:
> > @Madhu-1 yes, I have tested this delete path and it is working.
> > Reg, adding E2E, unfortunately I dont see a straight way here, that said, the volHandle and Image name has to be manipulated either by following our normal PVC creation with proper SC, or the SC details and other attributes in the PVC/PV spec need to be changed after creation if its static PV workflow, both looks to be difficult or not possible to achieve.
> 
> I think it can be done easily. Can you make use of the `getStaticPV` function
> 
>     * Create static PVC and static PV
> 
>     * Create an application pod and verify you are able to get the pod running
> 
> 
> The above steps are already done in static migration tests
> 
> extra steps are below
> 
>     * when creating the static PV generate the volumeHandle use it
> 
>     * Change the reclaimPolicy from `Retain` to delete which will trigger rbd volume deletion.

I doubt this will work, the reason being the staticPV functions record a SC ( a dummy one) in the object?

humblec commented:
> > ```
> > * when creating the static PV generate the volumeHandle use it
> > 
> > * Change the reclaimPolicy from `Retain` to delete which will trigger rbd volume deletion.
> > ```
> 
> I doubt this will work, the reason being the staticPV functions record a SC ( a dummy one) in the object?

As discussed, I can create a SC and then record proper SC in PV spec  and give a try (eventhough its bit ugly :) ), lets see how that goes. :+1:

humblec commented:
> > > ```
> > > * when creating the static PV generate the volumeHandle use it
> > > 
> > > * Change the reclaimPolicy from `Retain` to delete which will trigger rbd volume deletion.
> > > ```
> > 
> > 
> > I doubt this will work, the reason being the staticPV functions record a SC ( a dummy one) in the object?
> 
> As discussed, I can create a SC and then record proper SC in PV spec and give a try (eventhough its bit ugly :) ), lets see how that goes. +1

The SC was filled https://github.com/ceph/ceph-csi/pull/2523/commits/77e8009ffaa615f47db2afc6be11348e4599ad4a#diff-fb1cee4ac1e01fbab0746d1a459dc370b11d272949985a858f7ef74f41d47b7cR349) and test deleted pvc and pv, but it does not trigger  the deletecall to the driver as suspected,.

@Madhu-1 please confirm, so that I can remove the last commit which is just e2e and get this refreshed.

Madhu-1 commented:
Yes you also need to change PV ReclaimPolicy to Delete at
https://github.com/ceph/ceph-csi/commit/77e8009ffaa615f47db2afc6be11348e4599ad4a#diff-fb1cee4ac1e01fbab0746d1a459dc370b11d272949985a858f7ef74f41d47b7cR381-R390

On Thu, Sep 23, 2021 at 5:23 PM Humble Devassy Chirammal <
***@***.***> wrote:

> * when creating the static PV generate the volumeHandle use it
>
> * Change the reclaimPolicy from `Retain` to delete which will trigger rbd volume deletion.
>
> I doubt this will work, the reason being the staticPV functions record a
> SC ( a dummy one) in the object?
>
> As discussed, I can create a SC and then record proper SC in PV spec and
> give a try (eventhough its bit ugly :) ), lets see how that goes. +1
>
> The SC was filled 77e8009
> #diff-fb1cee4ac1e01fbab0746d1a459dc370b11d272949985a858f7ef74f41d47b7cR349
> <https://github.com/ceph/ceph-csi/commit/77e8009ffaa615f47db2afc6be11348e4599ad4a#diff-fb1cee4ac1e01fbab0746d1a459dc370b11d272949985a858f7ef74f41d47b7cR349>)
> and test deleted pvc and pv, but it does not trigger the deletecall to the
> driver as suspected,.
>
> @Madhu-1 <https://github.com/Madhu-1> please confirm, so that I can
> remove the last commit which is just e2e and get this refreshed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph-csi/pull/2523#issuecomment-925739896>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKGNNYNDPV6G4QQLTUU5N3UDMIKBANCNFSM5EL3EILQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Madhu Rajanna
Senior Software Engineer

Red Hat Bangalore, India <https://www.redhat.com>

***@***.***    M: +91-9741133155
<https://red.ht/sig>

humblec commented:
> Yes you also need to change PV ReclaimPolicy to Delete at..

yeah, just noticed the version in CI run was different from the local commit, so refreshed the PR , hopefully it will go through. any way lets see.

humblec commented:
> > Yes you also need to change PV ReclaimPolicy to Delete at..
> 
> yeah, just noticed the version in CI run was different from the local commit, so refreshed the PR , hopefully it will go through. any way lets see.

@Madhu-1 even with the delete policy, the same result as previous run. ptal. once confirmed I will remove the e2e commit.

humblec commented:
@Madhu-1 @Yuggupta27 all comments are addressed. e2e also added, but looks like that has to be removed, ptal .  thanks.

yati1998 reviewed (commented):
Please do check out the golangci-lint failure.

humblec commented:
> Please do check out the golangci-lint failure.

@yati1998 both linter and go test failure is  on the e2e commit which is about to get removed (soon) after confirmation from  @Madhu-1 on this.... so kept it that way .

humblec commented:
> @Madhu-1 @Yuggupta27 all comments are addressed. e2e also added, but looks like that has to be removed, ptal . thanks.

Can you please review this ?

Madhu-1 commented:
> > > Yes you also need to change PV ReclaimPolicy to Delete at..
> > 
> > 
> > yeah, just noticed the version in CI run was different from the local commit, so refreshed the PR , hopefully it will go through. any way lets see.
> 
> @Madhu-1 even with the delete policy, the same result as previous run. ptal. once confirmed I will remove the e2e commit.

```
kubectl get pvc
NAME            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
cephfs-pvc-sc   Bound    static-pv                                  1Gi        RWO            rook-cephfs

kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS     CLAIM                   STORAGECLASS      REASON   AGE
static-pv                                  1Gi        RWO            Delete           Released   default/cephfs-pvc-sc   rook-cephfs                79s

kubectl delete pvc/cephfs-pvc-sc
persistentvolumeclaim "cephfs-pvc-sc" deleted


1 controller.go:1472] delete "static-pv": started
I0928 06:20:47.179040       1 connection.go:182] GRPC call: /csi.v1.Controller/DeleteVolume
I0928 06:20:47.179372       1 connection.go:183] GRPC request: {"secrets":"***stripped***","volume_id":"static-pv-handle"}
I0928 06:20:47.207441       1 connection.go:185] GRPC response: {}
I0928 06:20:47.207809       1 connection.go:186] GRPC error: rpc error: code = Internal desc = invalid VolumeID: error decoding volume ID (static-pv-handle): failed to decode CSI identifier, string underflow
E0928 06:20:47.208194       1 controller.go:1482] delete "static-pv": volume deletion failed: rpc error: code = Internal desc = invalid VolumeID: error decoding volume ID (static-pv-handle): failed to decode CSI identifier, string underflow
```

In my local testing I can see DeleteVolume Request for a static PV/PVC

humblec commented:
@Madhu-1 Added annotation hack too, hopefully it helps to simulate the situation. lets see..

humblec commented:
@Madhu-1 finally E2E is passing.  Thanks :+1:  ..  splitting of the commit is pending,  please take a final look ... will address commit split in next refresh..

Madhu-1 commented on e2e/pod.go:
```diff
@@ -374,6 +374,7 @@ func deletePod(name, ns string, c kubernetes.Interface, t int) error {
 	})
 }
 
+// nolint:unparam // currently skipNotFound is always false, this can change in the future
```
any idea why it's not found in current CI? if it's causing a CI issue it should be a separate commit.

humblec replied:
no idea , however in yesterday's patch refresh, CI started to complain about it.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -366,6 +366,59 @@ var _ = Describe("RBD", func() {
+				}
+				mon := strings.Join(mons, ",")
+				inClusterID := getMonsHash(mon)
+				customInfo := map[string]string{
+					"clusterID-1": "subvolgrp1",
+					// the monitors hash entry in the configmap.
+					inClusterID: "subvolgrp2",
+				}
```
IMO subvolgrp doesn't make sense for  RBD. please change this to something else

humblec replied:
Done.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -366,6 +366,59 @@ var _ = Describe("RBD", func() {
+				if err != nil {
+					e2elog.Failf("failed to create storageclass with error %v", err)
+				}
+				// restart csi pods for the configmap to take effect.
+				err = deletePodWithLabel("app in (ceph-csi-rbd, csi-rbdplugin, csi-rbdplugin-provisioner)",
+					cephCSINamespace, false)
+				if err != nil {
+					e2elog.Failf("failed to delete pods with labels with error %v", err)
+				}
+				// wait for csi pods to come up
+				err = waitForDaemonSets(rbdDaemonsetName, cephCSINamespace, f.ClientSet, deployTimeout)
+				if err != nil {
+					e2elog.Failf("timeout waiting for daemonset pods with error %v", err)
+				}
+				err = waitForDeploymentComplete(rbdDeploymentName, cephCSINamespace, f.ClientSet, deployTimeout)
+				if err != nil {
+					e2elog.Failf("timeout waiting for deployment to be in running state with error %v", err)
+				}
```
This can be moved to helper function and it can be reused in other places also.

humblec replied:
Done

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -53,6 +63,56 @@ func getStaticPV(
+		volumeMode := v1.PersistentVolumeFilesystem
+		pv.Spec.VolumeMode = &volumeMode
+	}
+	if pv.Annotations == nil {
```
This check is not required here as pv is a local object.

humblec replied:
Done

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -53,6 +63,56 @@ func getStaticPV(
 	return pv
 }
 
+func getMigrationStaticPV(
```
cant we reuse the `getStaticPV` and add/update what is required for migration instead of recreating the PV object?

humblec replied:
Done.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -53,6 +63,56 @@ func getStaticPV(
+						Name:      secretName,
+						Namespace: secretNS,
+					},
+					ControllerExpandSecretRef: &v1.SecretReference{
+						Name:      secretName,
+						Namespace: secretNS,
+					},
```
ControllerExpandSecretRef is required?

humblec replied:
can be removed. Done

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -366,6 +366,59 @@ var _ = Describe("RBD", func() {
+					e2elog.Failf("timeout waiting for deployment to be in running state with error %v", err)
+				}
+
+				err = validateRBDStaticMigrationPVDeletion(f, rawAppPath, true)
```
pass created sc as input parameter to the function and use it to create PV/PVC object

humblec replied:
Done.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -285,6 +345,105 @@ func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBloc
 	return err
 }
 
+func validateRBDStaticMigrationPVDeletion(f *framework.Framework, appPath string, isBlock bool) error {
```
move this to a new file migration.go in e2e?

humblec replied:
Done

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -285,6 +345,105 @@ func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBloc
+	cmd := fmt.Sprintf(
+		"rbd create %s --size=%d --image-feature=layering %s",
+		rbdImageName,
+		4096,
```
use  `staticPVSize` size here?

humblec replied:
Done.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -285,6 +345,105 @@ func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBloc
+		4096,
+		rbdOptions(defaultRBDPool))
+
+	_, e, err := execCommandInToolBoxPod(f, cmd, rookNamespace)
```
e variable should be stdErr?

humblec replied:
Done.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -285,6 +345,105 @@ func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBloc
+		rbdNodePluginSecretName,
+		cephCSINamespace,
+		sc,
+		"rbd.csi.ceph.com",
```
define const for driverName?

humblec replied:
Done.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -285,6 +345,105 @@ func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBloc
+		return err
+	}
+
+	err = deletePod(app.Name, app.Namespace, f.ClientSet, deployTimeout)
+	if err != nil {
+		return err
+	}
+	err = c.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(context.TODO(), pvc.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to delete pvc: %w", err)
+	}
```
use `deletePVCAndApp` which validates the PVC and PV deletion also

humblec replied:
Done.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -285,6 +345,105 @@ func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBloc
+	return err
+}
+
+func composeIntreeMigHandle(mons, rbdImageName string) string {
```
please add a comment about this function.

humblec replied:
Done.

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -285,6 +345,105 @@ func validateRBDStaticMigrationPV(f *framework.Framework, appPath string, isBloc
+	vhSlice := []string{migIdentifier, monsField, imageField, poolField}
+	volHandle := strings.Join(vhSlice, "_")
+
+	return volHandle
```
return `strings.Join(vhSlice, "_")` no need to create extra  variable.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +833,16 @@ func (cs *ControllerServer) DeleteVolume(
 	defer cs.OperationLocks.ReleaseDeleteLock(volumeID)
 
+	if isMigrationVolHandle(volumeID) {
+		log.DebugLog(ctx, "migration volume handle : %s and secrets: %v", volumeID, req.GetSecrets())
```
why we are logging secret here?

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +833,16 @@ func (cs *ControllerServer) DeleteVolume(
+		log.DebugLog(ctx, "migration volume handle : %s and secrets: %v", volumeID, req.GetSecrets())
+		err = parseAndDeleteMigratedVolume(ctx, volumeID, cr)
+		if err != nil {
+			return nil, err
```
parseAndDeleteMigratedVolume is not returning GRPC error code/message in all error cases. please make sure it returns proper GRPC errors or return a GRPC error from here

Madhu-1 commented on internal/rbd/errors.go:
```diff
@@ -34,4 +34,12 @@ var (
 	ErrMissingStash = errors.New("missing stash")
 	// ErrFlattenInProgress is returned when flatten is in progress for an image.
 	ErrFlattenInProgress = errors.New("flatten in progress")
+	// ErrMissingMonitorsInVolID is returned when monitor information is missing in migration vol Handle.
```
`vol Handle` need to be `volumeID` from CSI we receive request for volumeID

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,118 @@
+	"google.golang.org/grpc/status"
+)
+
+// isMigrationVolHandle validate the passed in volHash is a volume handle
```
volume handle needs to be replaced with volID. please take care of it in other places also

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,118 @@
+	for _, field := range handSlice[:migHandleSplitLength] {
+		switch {
+		case field == migIdentifier:
+			mh.identifier = true
```
are we making `identifier` anywhere for validation?

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,118 @@
+	rv.ClusterID = parsedMigHandle.clusterID
+	rv.Monitors, err = util.Mons(util.CsiConfigFile, rv.ClusterID)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to fetch monitors using clusterID: %s", rv.ClusterID)
```
please return error from here

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,118 @@
+	}
+	defer rv.Destroy()
+	// if connected , delete it
+	err = deleteImage(ctx, rv, cr)
```
imageNotFound case is already handled somewhere? am not seeing it anywhere

humblec replied:
@Madhu-1 getImageID() returns ErrImageNotFound if provided image is not found. thats what I was referring :)

Madhu-1 replied:
if it returns `ErrImageNotFound` same will be returned from the `parseAndDeleteMigratedVolume` function and that will be treated as the internalserver error in `DeleteVolume`?

humblec replied:
@Madhu-1 yeah, thats what I intended with this version. In one way the delete request is broken which matches with `internal` server error expectation.  do you want to change it to specifically treat `ErrImageNotFound` and return `success`?

Madhu-1 replied:
Not sure why it's intended can you please elaborate more on this? What if the delete request is sent to csi driver multiple times? All subsequent requests will start failing and pv will never gets deleted? Is it against the csi specifications?

humblec replied:
yeah, good to treat that specific case as success.. thanks

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,118 @@
+	// if connected , delete it
+	err = deleteImage(ctx, rv, cr)
+	if err != nil {
+		log.DebugLog(ctx, "failed to delete rv: %+v", rv)
```
remote this debug log

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,118 @@
+	// connect to the volume.
+	err = rv.Connect(cr)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to get connected to the rbd image : %s", rv.RbdImageName)
```
log the error message

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,118 @@
+	rv.ClusterID = parsedMigHandle.clusterID
+	rv.Monitors, err = util.Mons(util.CsiConfigFile, rv.ClusterID)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to fetch monitors using clusterID: %s", rv.ClusterID)
```
log error message

humblec replied:
Done.

Madhu-1 commented on internal/rbd/migration_test.go:
```diff
@@ -0,0 +1,165 @@
+func TestIsMigrationVolHandle(t *testing.T) {
+	t.Parallel()
+	volID := "wrong_volume_handle"
+
+	if isMigrationVolHandle(volID) {
+		t.Error("request is not a migration delete request")
+	}
+
+	volID = "mig_mons-b7f67366bb43f32e07d8a261a7840da9_image-e0b45b52-7e09-47d3-8f1b-806995fa4412_pool_replica_pool" // nolint:lll // migration handle
+	if !isMigrationVolHandle(volID) {
+		t.Error("request is a migration delete request")
+	}
+
+	volID = "mig_image-e0b45b52-7e09-47d3-8f1b-806995fa4412_pool_replica_pool"
+	if !isMigrationVolHandle(volID) {
+		t.Error("request is a migration delete request")
+	}
+
+	volID = "mig_image-e0b45b52-7e09-47d3-8f1b-806995fa4412"
+	if !isMigrationVolHandle(volID) {
+		t.Error("request is a migration delete request")
+	}
+	volID = "mig_"
+	if isMigrationVolHandle(volID) {
+		t.Error("request is not a migration delete request")
+	}
+
+	volID = "kubernetes-dynamic-pvc-e0b45b52-7e09-47d3-8f1b-806995fa4412"
+	if isMigrationVolHandle(volID) {
+		t.Error("request is not a migration delete request")
+	}
```
can this be rewritten with a simple for loop and a variable to check its migrationVolumeHandle or not?

Madhu-1 reviewed (commented):
commit message and PR description need to be updated to match new changes.

Madhu-1 commented:
Please reuse the same logic to add mon hash as clusterid and us the same in nodestage operator and remove the code to get clusterid from mon ip

humblec commented:
> Please reuse the same logic to add mon hash as clusterid and us the same in nodestage operator and remove the code to get clusterid from mon ip

@Madhu-1 I was thinking to address it in a follow up PR, wdyt?

Madhu-1 commented:
> > Please reuse the same logic to add mon hash as clusterid and us the same in nodestage operator and remove the code to get clusterid from mon ip
> 
> @Madhu-1 I was thinking to address it in a follow up PR, wdyt?

Am fine with that please mark it as todo and also update the e2e for same as we previously added in other ip for migration pvc

humblec commented:
@Madhu-1 all the comments from last review have been addressed. ptal . thanks

humblec committed:
e2e: change createCustomConfigmap to be more general

createCustomConfigmap helps to create a custom cluster entry in
the configmap, however this was coupled with subvolumegroup filling
in the cluster configuration. This commit helps to make it more
general and the subvolumegroup filling is controlled now with a flag

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
e2e: deletePodWithLabel fails on unparam linter

this commit address the unparam linter error on deletePodWithLabel
function.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
e2e: add test for migration volID detection and delete of image

This commit add test for migration delete volID detection scenario
by passing a custom volID and with the entries in configmap changed
to simulate the situation. The staticPV function also changed its
accept the annotation map which make it more general usage.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: detect migration volID in DeleteVolume() and delete rbd image

This commit adds the logic to detect a passed in volumeID
is a migrated volume ID and if yes, the driver connect to the
backend cluster and clean/delete the image. The logic
only applied if its a migration volume ID. The migration volume ID
carry the information like mons, pool and image name which is
good enough for the driver to identify and connect to the backend
cluster for its operations.

migration volID format:
<mig>_mons-<monsHash>_image-<imageUID>_<poolHash>

Details on the hash values:

* MonsHash: this carry a hash value (md5sum) which will be acted as the
`clusterID` for the operations in this context.

* ImageUID: this is the unique UUID generated by kubernetes for the created
volume.

* PoolHash: this is an encoded string of pool name.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 commented on e2e/configmap.go:
```diff
@@ -76,7 +76,11 @@ func createConfigMap(pluginPath string, c kubernetes.Interface, f *framework.Fra
+func createCustomConfigMap(
+	c kubernetes.Interface,
+	pluginPath string,
+	clusterInfoMap map[string]string,
+	skipSubvolFill bool) error {
```
This change is not flexible either clusterInfoMap should be a map[string]map[string]string{}  with key as clusterID and the other map contains the subvolumeGroup and radosnamespace. 
 sample code

```suggestion
clusterInfo:=map[string]map[string]string{}
clusterInfo[clusterID1] = map[string]string{}
clusterInfo[clusterID1] ["subvolumeGrop"]="subvolumegroup1"
clusterInfo[clusterID1] ["radosnamepsace"]="radosns1"

clusterInfo[clusterID2] = map[string]string{}
clusterInfo[clusterID2] ["subvolumeGrop"]="subvolumegroup2"
clusterInfo[clusterID2] ["radosnamepsace"]="radosns2"
```

Based on the clusterID fill clusterID details if the map contains the subvolume update subvolumegroup for the clusterID and same can be done for radosnamespace also

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -366,6 +366,48 @@ var _ = Describe("RBD", func() {
+				mon := strings.Join(mons, ",")
+				inClusterID := getMonsHash(mon)
+				customInfo := map[string]string{
+					"clusterID-1": "",
```
clusterID-1 is a dummy and not needed?

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -833,6 +833,16 @@ func (cs *ControllerServer) DeleteVolume(
+		log.DebugLog(ctx, "migration volume ID : %s", volumeID)
+		err = parseAndDeleteMigratedVolume(ctx, volumeID, cr)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
```
`status.Errorf` should be `status.Error` as no formating is done

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,116 @@
+	rv := &rbdVolume{}
+
+	// fill details to rv struct from parsed migration handle
+	rv.RbdImageName = parsedMigHandle.imageName
+	rv.Pool = parsedMigHandle.poolName
+	rv.ClusterID = parsedMigHandle.clusterID
+	rv.Monitors, err = util.Mons(util.CsiConfigFile, rv.ClusterID)
```
parseMigrationVolID should be the method of `rbdVolume`? if we do that it avoids creation/maintaining an extra structure.

humblec replied:
it is better to be kept outside of rv, mainly bceause it can be independently used to call and also just retrive the migration handle specific data which is one step ahead of rv parsing.

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,116 @@
+	rv.ClusterID = parsedMigHandle.clusterID
+	rv.Monitors, err = util.Mons(util.CsiConfigFile, rv.ClusterID)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to fetch monitors using clusterID: %s, err:%v", rv.ClusterID, err)
```
`err:%v`  to `err: %v`

Madhu-1 reviewed (dismissed):
Need few changes else LGTM

humblec commented:
>if it returns ErrImageNotFound same will be returned from the parseAndDeleteMigratedVolume function and that will be treated as the internalserver error in DeleteVolume?

@Madhu-1  thats what I intended with this version,  is it the suggestion to return `success` if the error was `image not found` ?

humblec commented:
> Need few changes else LGTM

@Madhu-1 all the comments are addressed. ptal at final version. Thanks :+1:

Yuggupta27 commented on e2e/rbd_helper.go:
```diff
@@ -974,3 +974,22 @@ func waitToRemoveImagesFromTrash(f *framework.Framework, poolName string, t int)
 	return err
 }
+
+func recreateCSIRBDPods(f *framework.Framework) error {
```
Instead of keeping this function just for RBD, we can make it available even for cephfs, just by just passing the label as a parameter; so that in future we don't have to rewrite the same function for cephfs. Also, then we can restart any pod(not just csi) with labels. The function can then be renamed `recreatePods()`. What do you think?

humblec replied:
@Yuggupta27 can be done that improvement if needed, I can do that as a followup , marking for todo

Yuggupta27 replied:
Thanks

Yuggupta27 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,116 @@
+	"github.com/ceph/ceph-csi/internal/util/log"
+)
+
+// isMigrationVolID validate the passed in volID is a volumeID
+// of migrated volume.
```
```
// isMigrationVolID validates if the passed in volID is a volumeID
// of a migrated volume.
```

humblec replied:
Done. thanks.

Yuggupta27 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,116 @@
+// isMigrationVolID validate the passed in volID is a volumeID
+// of migrated volume.
+func isMigrationVolID(volHash string) bool {
+	return strings.Contains(volHash, migIdentifier) &&
```
Since it is an identifier, is the `migIdentifier` is not enough for validating if the volume is in context to migration? It might reduce 2 checks

humblec replied:
the other checks are required to be sure.

Yuggupta27 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,116 @@
+		strings.Contains(volHash, migImageNamePrefix) && strings.Contains(volHash, migMonPrefix)
+}
+
+// parseMigrationVolID decode the volume ID and generate a migrationVolID
```
decode -> decodes, generate -> generates

humblec replied:
Done.. thanks

Yuggupta27 commented on internal/rbd/migration.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// parseMigrationVolID decode the volume ID and generate a migrationVolID
+// struct which consists of version, mon, image name , pool and clusterID information.
```
`image name , pool` -> `image name, pool`

humblec replied:
Done. thanks.

humblec commented:
@Mergifyio rebase

humblec commented:
/retest  ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest ci/centos/upgrade-tests-cephfs

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest  ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22
